File: 0010-regular-expression-matching/0010-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        # Create a memoization dictionary
        memo = {}

        def posMatch(i: int, j: int) -> bool:
            # Check if the result is already computed
            if (i, j) in memo:
                return memo[(i, j)]

            if i == len(s):
                if j == len(p):
                    memo[(i, j)] = True
                elif j + 1 < len(p) and p[j + 1] == '*':
                    memo[(i, j)] = posMatch(i, j + 2)
                else:
                    memo[(i, j)] = False
                return memo[(i, j)]

            if j == len(p):
                memo[(i, j)] = False
                return False

            if s[i] == p[j] or p[j] == '.':
                if j + 1 < len(p) and p[j + 1] == '*':
                    memo[(i, j)] = posMatch(i + 1, j) or posMatch(i, j + 2)
                else:
                    memo[(i, j)] = posMatch(i + 1, j + 1)
            else:
                if j + 1 < len(p) and p[j + 1] == '*':
                    memo[(i, j)] = posMatch(i, j + 2)
                else:
                    memo[(i, j)] = False

            return memo[(i, j)]

        # Start the recursive matching from (0, 0)
        return posMatch(0, 0)

File: 0010-regular-expression-matching/0010-regular-expression-matching.cpp
#include <string>

using namespace std;

class Solution {
private:
    bool posMatch(const string& s, const string& p, int i, int j) {
        if (i == s.length()) {
            if (j == p.length()) {
                return true;
            } else if (j + 1 < p.length() && p[j + 1] == '*') {
                return posMatch(s, p, i, j + 2);
            } else {
                return false;
            }
        }
        
        if (j == p.length()) {
            return false;
        }
        
        if (s[i] == p[j] || p[j] == '.') {
            if (j + 1 < p.length() && p[j + 1] == '*') {
                return posMatch(s, p, i + 1, j) || posMatch(s, p, i, j + 2);
            } else {
                return posMatch(s, p, i + 1, j + 1);
            }
        } else {
            if (j + 1 < p.length() && p[j + 1] == '*') {
                return posMatch(s, p, i, j + 2);
            }
            return false;
        }
    }
    
public:
    bool isMatch(string s, string p) {
        return posMatch(s, p, 0, 0);
    }
};
